### DISCUSSION TITLE
Windows usage via PIP

### DISCUSSION THREAD

**Discussion by Unichord** [INTI-CMNB/KiBot#325] [Other platforms]
Hello,
I'm trying to setup KiBot on Windows 10 via PIP. I installed it pretty seamlessly, but apparently I did/I'm doing something wrong, since the actual command doesn't work as expected and I don't really have experience with Python/PIP and I can't figure out what's wrong.
I'm running the command via the KiCAD 6.0 Command Prompt as Administrator (I also tried as normal user), I already installed all the dependencies, I think, and I also added environment variables.

Also kibot-check doesn't work, it's not recognized as a command, I don't know why. Maybe I'm using it wrong?

I'd like to try once more to make this work, otherwise I'll go with the WSL method. Any suggestions? Thanks in advance.

Here is the command I run: [https://pastebin.com/LnXLPjha](https://pastebin.com/LnXLPjha)

I now noticed that KiBot creates this empty folder, I don't know if it's an useful piece of information: C:\Users\Andrea\Documents\Progetti\CLE\FMFR1_PROVA_6.0\Andrea\.local\share\kibot\bin

**Comment by set-soft** [comment]
Hi @Unichord !
Currently KiBot needs WSL or docker to run on Windows.
As you can see KiBot  does a lot of things on Windows, but needs more work to be usable. Also: even if we make it usable some stuff won't work. Most things done by KiAuto will still need WSL.

**Reply by Unichord** [reply]
Hi @set-soft, thanks for the answer (and for KiBot obviously).
I'll try to give WSL a go, I'll follow the guide in this subsection.

**Reply by set-soft** [reply]
Ok, please comment about any thing that needs to be adjusted in the WSL description. Is new and each system has its own particularities.

**Comment by edlongman** [comment]
I have been able to somewhat successfully run KiBot in Windows - and it's extremally useful, thank you! 

I follow the following process/workarounds
  - Setup environment with the KiCad Python 3.9 in the path. 
  - Install with `pip install kibot --no-compile`
  - I run `kibot` within Git Bash to use git tools for `set_text_variables`
  - Apply this attached patch to `pre_set_text_variables.py`
```
diff --git a/kibot/pre_set_text_variables.py b/kibot/pre_set_text_variables.py
old mode 100644
new mode 100755
index 8043b16..9060376
--- a/pre_set_text_variables.py
+++ b/pre_set_text_variables.py
@@ -119,11 +119,7 @@ class Set_Text_Variables(BasePreFlight):  # noqa: F821
             text = r.text
             if not text and r.command:
                 command = r.command
-                if re_git.search(command):
-                    git_command = self.ensure_tool('git')
-                    command = re_git.sub(r'\1'+git_command+' ', command)
-                cmd = ['/bin/bash', '-c', command]
-                result = run(cmd, stdout=PIPE, stderr=PIPE, universal_newlines=True)
+                result = run(command, stdout=PIPE, stderr=PIPE, universal_newlines=True, shell=True)
                 if result.returncode:
                     logger.error('Failed to execute:\n{}\nreturn code {}'.format(r.command, result.returncode))
                     sys.exit(FAILED_EXECUTE)
```
 - In `kibot.yaml` use `/../` before every dir output to make relative to local directory. (Avoids sub directories of `C_/Users/........./Documents/KiCad/...` being added

Obviously I still can't use KiAuto outputs, but gerbers, pos, drills and BOM works well.

**Reply by set-soft** [reply]
Hi @edlongman !
Thanks for sharing your experience.

Did you use the "KiCad Command Prompt"?  (Explained in the [Pcbdraw installation](https://github.com/yaqwsx/PcbDraw/blob/master/doc/installation.md))

> Apply this attached patch to pre_set_text_variables.py

Can you find what is failing here? My guess is that `/bin/bash` isn't valid, but may be is worst.
You can add a line with `logger.error(cmd)` after the cmd assignment and before the run sentence. This will show what is the code trying to run. Perhaps we can just adjust the shell name.

> In kibot.yaml use /../ before ...

We could try to fix it. I think the problem is in the `%X` variable expansion and in the absolute path detection. Can you run KiBot using `-vvv` and look for the `C_/Users/........./Documents/KiCad/...` expansions? 

> Obviously I still can't use KiAuto outputs, but gerbers, pos, drills and BOM works well.

In addition, any output needing external tools like ghostscript, rsvg-convert, ImageMagick, etc. will fail if these tools aren't installed. And the *automatic* installation will fail.

We could make most things work in native mode, and let WSL only for KiAuto. From what I read in the KiCad Forum various things done by KiAuto will become easier in v7, they are introducing a `kicad-cli` tool which can do some stuff from the command line.

**Reply by edlongman** [reply]
Hi @set-soft,




> Did you use the "KiCad Command Prompt"? (Explained in the [Pcbdraw installation](https://github.com/yaqwsx/PcbDraw/blob/master/doc/installation.md))

I used the git bash command prompt so that I can have access to tools like `sed` and `grep`.

> > Apply this attached patch to pre_set_text_variables.py
> 
> Can you find what is failing here? My guess is that `/bin/bash` isn't valid, but may be is worst. You can add a line with `logger.error(cmd)` after the cmd assignment and before the run sentence. This will show what is the code trying to run. Perhaps we can just adjust the shell name.

Yes, what is failing is the Windows directory separators "\\" being interpreted as escape chars. From adding `logger.error(git_command)` and `logger.error(command)` in before pre_set_text_variables.py:124 I see:
```
git_command = "C:\Users\...\AppData\Local\Programs\Git\mingw64\bin\git.EXE"
command = "git describe --exact-match --tag"
```
Resulting in error:
```
...
  File "C:\Users\...\Documents\KiCad\6.0\3rdparty\Python39\site-packages\kibot\pre_base.py", line 78, in run_enabled
    v.apply()
  File "C:\Users\...\Documents\KiCad\6.0\3rdparty\Python39\site-packages\kibot\pre_set_text_variables.py", line 126, in app
ly
    command = re_git.sub(r'\1'+git_command+' ', command)
  File "C:\Program Files\KiCad\6.0\bin\lib\re.py", line 327, in _subx
    template = _compile_repl(template, pattern)
  File "C:\Program Files\KiCad\6.0\bin\lib\re.py", line 318, in _compile_repl
    return sre_parse.parse_template(repl, pattern)
  File "C:\Program Files\KiCad\6.0\bin\lib\sre_parse.py", line 1054, in parse_template
    raise s.error('bad escape %s' % this, len(this))
re.error: bad escape \U at position 4
```

> > In kibot.yaml use /../ before ...
> 
> We could try to fix it. I think the problem is in the `%X` variable expansion and in the absolute path detection. Can you run KiBot using `-vvv` and look for the `C_/Users/........./Documents/KiCad/...` expansions?
> 
From running `kibot --vvv` the following relevant lines are provided:
```
DEBUG:Board loaded (kibot - __init__.py:1434)
DEBUG:Using global `output`=`%f-%i%I%v.%x` (kibot - __init__.py:1434)
- '' (gerber) [gerber]
DEBUG:Output destination: C:\Users\...\Documents\Projects\Board\C_\Users\...\Documents\Projects\Board\manufacturing (kibot - __init__.py:1434)
DEBUG:Configuring plot controller for output (kibot - __init__.py:1434)


```


Sorry, A lot of things to address there! I hope that helps and let's continue the conversation.

**Reply by set-soft** [reply]
Hi @edlongman !

> Yes, what is failing is the Windows directory separators "\" being interpreted as escape chars. From adding `logger.error(git_command)` and `logger.error(command)` in before pre_set_text_variables.py:124 I see:

Thanks! Totally unexpected to me! The following patch should fix it: fdd38d176ad90c6c211b9dde3874483afd1739e7

> From running kibot --vvv the following relevant lines are provided:

Try adding one more `v` and look for messages like this `Expanding `{}` in {} context for {} parent: {}` and their result.
But I suggest trying the following patch first: 75ebc7a6776e5187dc5285f87094c8d7087e1340

This should avoid the `C_` problem.

Note: both patches are part of the `dev` branch.

**Reply by edlongman** [reply]
> Thanks! Totally unexpected to me! The following patch should fix it: fdd38d176ad90c6c211b9dde3874483afd1739e7

Regarding the patch, might be worth using `re.escape(...)` Instead of replace as other nasties like `$1` could appear in the directory listing!


> Try adding one more `v` and look for messages like this `Expanding `{}` in {} context for {} parent: {}` and their result.
> But I suggest trying the following patch first: 75ebc7a6776e5187dc5285f87094c8d7087e1340
> 
> This should avoid the `C_` problem.
> 
> Note: both patches are part of the `dev` branch.

I will try later, but I would note on this fix that you can still do the replacement of `\\` with `/` in windows.

Also, path sanitization may best be done with an external package like [pathvalidator](https://pypi.org/project/pathvalidate/).

**Reply by set-soft** [reply]
> Regarding the patch, might be worth using `re.escape(...)` Instead of replace as other nasties like `$1` could appear in the directory listing!

It was the first thing I tried but doesn't work, the Python docs explicitly say not to use `re.escape` for `re.sub`, and says you must just escape the backslash.

> I will try later, but I would note on this fix that you can still do the replacement of \\ with / in windows.

I know, but the Python used by KiCad insists on using backslashes, I guess this is MinGW using Microsoft C runtime. 

> Also, path sanitization may best be done with an external package like pathvalidator.

I try to avoid rare dependencies in the core of KiBot, but will have it mind, thanks.

**Reply by edlongman** [reply]
Hi @set-soft,
I've got around to testing it and running v1.5.1 (which I think has those changes in it) is at least a partial fix but getting a slightly different error from the following yaml:

```
preflight:
  set_text_variables:
  # Issue from latest git tag or show draft if no tag.
  - name: "ISSUE"
    command: "git describe --exact-match --tag"
```

and the errors
```
DEBUG:Starting tool check for Git (kibot - __init__.py:1434)
DEBUG:- Checking binary tool Git (kibot - log.py:124)
DEBUG:- Looking for tool `git` at system level (kibot - log.py:124)
DEBUG:- Checking version for `C:\Users\...\Git\mingw64\bin\git.EXE` (kibot - log.py:124)
DEBUG:- No particular version needed (kibot - log.py:124)
DEBUG:- Running ['C:\\Users\\...\\Git\\mingw64\\bin\\git.EXE', '--version'] (kibot - log.py:124)
DEBUG:- Looking for version in `git version 2.38.1.windows.1` (kibot - log.py:124)
DEBUG:- Found version (2, 38, 1, 0) (kibot - log.py:124)
DEBUG:- Returning `C:\Users\...\Git\mingw64\bin\git.EXE` (kibot - __init__.py:1434)
WARNING:Trying git command: ['/bin/bash', '-c', 'C:\\Users\\...\\Git\\mingw64\\bin\\git.EXE describe --exact-match --tag'] (kibot - log.py:102)
Traceback (most recent call last):
  File "C:\Program Files\KiCad\6.0\bin\Lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  ...
  File "C:\Users\...\KiCad\6.0\3rdparty\Python39\site-packages\kibot\pre_base.py", line 78, in run_enabled
    v.apply()
  File "C:\Users\...\KiCad\6.0\3rdparty\Python39\site-packages\kibot\pre_set_text_variables.py", line 126, in apply
    result = run(cmd, stdout=PIPE, stderr=PIPE, universal_newlines=True)
  File "C:\Program Files\KiCad\6.0\bin\Lib\subprocess.py", line 505, in run
    with Popen(*popenargs, **kwargs) as process:
  File "C:\Program Files\KiCad\6.0\bin\Lib\subprocess.py", line 951, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\Program Files\KiCad\6.0\bin\Lib\subprocess.py", line 1420, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
FileNotFoundError: [WinError 2] The system cannot find the file specified

```

And a new error to do with `/dev/null` Is this to do with deleting files a Unix way?
```
DEBUG:Outputs after sorting: [<kibot.out_gerber.Gerber object at 0x000001C409DA2B20>, <kibot.out_gerb_drill.Gerb_Drill object at 0x000
001C409DADC10>, <kibot.out_report.Report object at 0x000001C409DADA60>, <kibot.out_bom.BoM object at 0x000001C409DAD580>, <kibot.out_p
osition.Position object at 0x000001C409DAD7F0>, <kibot.out_pdf_sch_print.PDF_SCH_Print object at 0x000001C409DAD640>, <kibot.out_step.
STEP object at 0x000001C409DAD7C0>] (kibot - __init__.py:1434)
ERROR:Error loading PCB file. Corrupted? (kibot - __init__.py:1475)
ERROR:[Errno 2] No such file or directory: '/dev/null' (kibot - __init__.py:1475)
```

Worked around this by disabling `devnull` in `hide_stderr()` in misc.py and now it produces manufacturing gerbers well.

**Reply by set-soft** [reply]
Hi @edlongman !

Thanks for testing it!

The first error is related to the assumption that bash is part of the system. The 11773e920da9d0523e9de1e1f2dd4d86b2f442cb patch fixes it, now we look for Bash and inform if missing.
The second was related to assume that */dev/null* is there. Now we ask Python the name of the OS specific file, patch c1f7b8dcff98664efbeb35de3ff7ad2befd25b1e
Both patches are applied to the `dev` branch. It currently reports version 1.5.2.

**Reply by set-soft** [reply]
> Is this to do with deleting files a Unix way?

Not, just a mechanism to discard things. KiCad reports some confusing errors (from the wxWidgets lib) that are harmless and we won't to avoid showing them to the user. So we temporally suppress KiCad's messages. In my tests this only worked using very low level stuff, this is because the messages comes from C++ in KiCad, not the Python API.

**Reply by edlongman** [reply]
Works very nicely thanks! I'll keep testing and hopefully KiCad 7.0 will provide the necessary command interfaces for PDF and step outputs available.

KiBot seems unable to overwrite outputs for some reason so I have to manually delete the exports currently, maybe this is intended behaviour though?

```
Traceback (most recent call last):
  File "C:\Program Files\KiCad\6.0\bin\Lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Program Files\KiCad\6.0\bin\Lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
...
  File "C:\Users\...\KiCad\6.0\3rdparty\Python39\site-packages\kibot\kiplot.py", line 415, in run_output
    out.run(get_output_dir(out.dir, out))
  File "C:\Users\...\KiCad\6.0\3rdparty\Python39\site-packages\kibot\out_base.py", line 189, in run
    self.options.run(self.expand_filename(output_dir, output))
  File "C:\Users\...\KiCad\6.0\3rdparty\Python39\site-packages\kibot\out_any_drill.py", line 180, in run
    os.rename(k_f, f)
FileExistsError: [WinError 183] Cannot create a file when that file already exists: 'C:\\Users\\...\\KiCad Project\\manufacturing\\Part-PTH-drl.gbr' -> 'C:\\Users\\...\\KiCad Project\\manufacturing\\Part-PTH_drill.gbr'
```

**Reply by set-soft** [reply]
Hi @edlongman !

> Works very nicely thanks! I'll keep testing and hopefully KiCad 7.0 will provide the necessary command interfaces for PDF and step outputs available.

From what I saw in the forum only a few operations will be covered by *kicad-cli*, but the good thing is that they are moving in the right direction. They rejected *cli* patches for years.
 
> KiBot seems unable to overwrite outputs for some reason so I have to manually delete the exports currently, maybe this is intended behaviour though?

This seems to be a limitation of the Windows filesystems (rename is not atomic). So now I'm using `os.replace` instead. The ebabc0b37ba0f37db413611ade201e845a42235c patch should solve it.
Funny thing: I explicitly did the check for back-ups, but not in all the other uses.
